SWIPs/swip-23-kademlia.md

```diff
@@ -0,0 +1,238 @@
+---
+SWIP: 23
+title: Swarm node implementer spec - 
+author: Louis Holbrook <dev@holbrook.no> (https://holbrook.no)
+status: Draft
+type: Track Specs
+created: 2019-09-05
+---
+
+## Abstract
+
+This SWIP is a part of a general specification of a Swarm node. The SWIP system is used for convenience only.
+
+## Motivation
+
+Enable other implementations of the swarm node.
+
+## Backwards Compatibility
+
+N/A
+
+## Test Cases
+
+N/A
+
+## Implementation
+
+N/A
+
+## Copyright
+
+Copyright and related rights waived via CC0
+
+## Specification
+
+The Kademlia is both a database (DHT) of known and connected peers on
+the network, as well as a component for routing and forwarding content
+on the network.
+
+### Records
+
+A peer record *ALWAYS* contains the `Address Pair` and the
+`Capabilities` of the peer. This data structure is known as the
+`BzzAddress`. It serializes as follows:
+
+    BZZADDRESS = LIST(OVERLAYADDRESS UNDERLAYADDRESS CAPABILITIES)
+
+### Connectivity
+
+#### Address matching
+
+For two addresses to match each other, their byte values must be
+identical.
+
+For certain purposes in Swarm *partial addresses* are used. A partial
+address is an address of less than 32 bytes. The bytes of a partial
+address always represent the most significant bytes of the full address
+that the partial address represents. A partial address matches a full
+address if all the bytes in the partial address are identical to the
+corresponding bytes in the full address.
+
+#### Distance and proximity
+
+In Swarm, differences between addresses are described in terms of
+*Distance* and *Proximity*.
+
+The *Distance* between two addresses is the value of the big-endian
```

Comment:
This seems like math in plain text, something like:

The *distance* between two addresses A and B is defined as follows:

`distance(a,b) = f(xor(a,b))`

where f (or whatever better name) is <explain>

Reply:
e.g. https://swarm-guide.readthedocs.io/en/latest/architecture.html#overlay-network